pkg/install/resources.go

```diff
@@ -0,0 +1,161 @@
+
+import (
+	"fmt"
+
+	appsv1beta1 "k8s.io/api/apps/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1beta1 "k8s.io/api/rbac/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type Option func(*config)
+
+type config struct {
+	container string
+	tag       string
+}
+
+func (c *config) image() string {
+	return fmt.Sprintf("%s:%s", c.container, c.tag)
+}
+
+func WithContainer(container string) Option {
+	return func(c *config) {
+		c.container = container
+	}
+}
+
+func WithTag(tag string) Option {
+	return func(c *config) {
+		c.tag = tag
+	}
+}
+
+func labels() map[string]string {
+	return map[string]string{
+		"component": "ark",
+	}
+}
+
+func objectMeta(name, namespace string) metav1.ObjectMeta {
+	return metav1.ObjectMeta{
+		Name:      name,
+		Namespace: namespace,
+		Labels:    labels(),
+	}
+}
+
+func ServiceAccount(namespace string) *corev1.ServiceAccount {
+	return &corev1.ServiceAccount{
+		ObjectMeta: objectMeta("ark", namespace),
+	}
+}
+
+func ClusterRoleBinding(namespace string) *rbacv1beta1.ClusterRoleBinding {
+	return &rbacv1beta1.ClusterRoleBinding{
+		ObjectMeta: objectMeta("ark", namespace),
+		Subjects: []rbacv1beta1.Subject{
+			{
+				Kind:      "ServiceAccount",
+				Namespace: namespace,
+				Name:      "ark",
+			},
+		},
+		RoleRef: rbacv1beta1.RoleRef{
+			Kind:     "ClusterRole",
+			Name:     "cluster-admin",
+			APIGroup: "rbac.authorization.k8s.io",
+		},
+	}
+}
+
+func Deployment(namespace string, opts ...Option) *appsv1beta1.Deployment {
+	c := &config{
+		container: "gcr.io/heptio-images/ark",
+		tag:       "latest",
+	}
+
+	for _, opt := range opts {
+		opt(c)
+	}
+
+	pullPolicy := corev1.PullIfNotPresent
+	if c.tag == "latest" {
+		pullPolicy = corev1.PullAlways
+	}
+
+	return &appsv1beta1.Deployment{
+		ObjectMeta: objectMeta("ark", namespace),
+		Spec: appsv1beta1.DeploymentSpec{
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"component": "ark",
+					},
+				},
+				Spec: corev1.PodSpec{
+					RestartPolicy:      corev1.RestartPolicyAlways,
+					ServiceAccountName: "ark",
+					Containers: []corev1.Container{
+						{
+							Name:            "ark",
+							Image:           c.image(),
+							ImagePullPolicy: pullPolicy,
+							Command: []string{
+								"/ark",
+							},
+							Args: []string{
+								"server",
+							},
+							VolumeMounts: []corev1.VolumeMount{
+								{
+									Name:      "cloud-credentials",
+									MountPath: "/credentials",
+								},
+								{
+									Name:      "plugins",
+									MountPath: "/plugins",
+								},
+							},
+							Env: []corev1.EnvVar{
+								{
+									Name:  "GOOGLE_APPLICATION_CREDENTIALS",
+									Value: "/credentials/cloud",
+								},
+								{
+									Name:  "AWS_SHARED_CREDENTIALS_FILE",
+									Value: "/credentials/cloud",
+								},
+							},
+						},
+					},
+					Volumes: []corev1.Volume{
+						{
+							Name: "cloud-credentials",
```

Comment:
Some deployments don't require this volume (e.g. AWS using the node's IAM instance profile credentials). Maybe we should make it a `deploymentOption`?

Reply:
is there an alternative env that has to be set if this is the case?

Reply:
For that particular case, no.